### PR TITLE
Refactor Restore flow

### PR DESCRIPTION
**Pull request by erdemyerebasmaz**
This PR is part of #1183

It introduces AsyncActions to BackupBloc and makes InitialWalkthroughPage responsible from navigation.

All handler logic introduced in #1183 is removed and replaced with ListSnapshots & RestoreBackup async actions.
This also allowed us to remove all the stream/sinks used for communication in these flows.

### Changelist:
- Use ListSnapshots & RestoreBackup AsyncActions on restore UI flows 2795c09
- Handle all restore flow navigation on InitialWalkthroughPage b554858
- Move models on BackupBloc to backup_model.dart 38d14dbee2877af5ebda80a1e727c081111a30e2
- Add named constructor & copyWith functionality to BackupState 7e4cb6e
- Tidy up _restoreBackup logic for readability using await d0d9341

### Todo: 
- Removal of EasyLoading plugin (?)

### CODE REVIEW

File: lib/bloc/backup/backup_bloc.dart [hunk 1]
```diff
@@ -331,6 +319,78 @@ class BackupBloc {
         encryptionKey?.key, encryptionKey?.type);
   }
 
+  Future _listSnapshots(ListSnapshots action) async {
+    String backups = await _breezLib.getAvailableBackups().catchError(
```
Comment (roeierez) [inline lib/bloc/backup/backup_bloc.dart:323]: When using this pattern: await  _breezLib.getAvailableBackups().catchError(...) won't it return with success? It seems as the code will continue after the catch to line 343. Or am I wrong?

File: lib/bloc/backup/backup_bloc.dart [hunk 2]
```diff
@@ -331,6 +319,78 @@ class BackupBloc {
...
+    await _breezLib
+        .restore(
+          action.restoreRequest.snapshot.nodeID,
+          action.restoreRequest.encryptionKey.key,
+        )
+        .catchError(
+          (error) => action.resolveError(error),
+        );
+
+    await _restoreAppData().catchError((error) {
```
Comment (roeierez) [inline lib/bloc/backup/backup_bloc.dart:377]: same here

**Review by roeierez — CHANGES_REQUESTED** [review]
Looks good!

File: lib/bloc/backup/backup_bloc.dart (hunk 1 above)
Comment (erdemyerebasmaz) [reply lib/bloc/backup/backup_bloc.dart:323]: Yes, we're missing return calls after resolving with error.

Comment (erdemyerebasmaz) [reply lib/bloc/backup/backup_bloc.dart:323]: I will be debugging & addressing all error handling issues.

**Comment by erdemyerebasmaz** [comment]
# Restore Flow Error Handling Checklist

## SelectBackupProviderDialog
- ### Apple iCloud

- ### Google Drive
  - [ ] Reject Permission: SignInFailed
  - [ ] Cancel after giving permission: SignInFailed

- ### Remote Auth
  - [ ] Error authenticating
    - #### RemoteServerAuthPage
      - [ ] // TODO

- ### After Selecting a provider
    - [ ] Error on UpdateBackupSettings action
    - [ ] Error on ListSnapshots action
    - [ ] No available snapshots on account 
    
## RestoreDialog
- ### If snapshot is encrypted
  -  #### Using Mnemonic Seed
      - [ ] Error on SaveBackupKey action
      - [ ] Error on UpdateBackupSettings action
      - [ ] Error getting Mnemonic Seed
        - #### EnterMnemonicsPage
          - [ ] 12 Words seed
            - [ ] // TODO
          - [ ] 24 Words seed (_Backwards Compatibility_)
            - [ ] // TODO
    - #### Using PIN Code(_Backwards Compatibility_)
      - [ ] // TODO
- ### If snapshot is not encrypted

## Completing Restore
- [ ] Error on RestoreBackup action
  - [ ] Error on clearing app data
  - [ ] Error on restoring node
  - [ ] Error on restoring app data
  - [ ] Error on updating last backup time
- [ ] Error on reloading database & pos items
- [ ] Error on refreshing user model

**Comment by erdemyerebasmaz** [comment]
### Changelist:
- Handle Insufficient Permission & InvalidCredentials error via extractExceptionMessage workaround a095662 6fd3efa
  - Move signIn out of getAvailableBackups
- Remember selected backup provider e429afc
- Use number of available snapshots when setting height of RestoreDialog content c9fca9f
- Add assertion check for SnapshotInfo's nodeId & modifiedTime that it can not be null 874cc92
  - Throw and error when snapshot list is null, empty or of an unknown type a9388f7
- [WIP] Tidy up RemoteServerAuthPage & remember entered credentials 63f3c3e
- Add try/catch blocks & generic error messages to BackupBloc methods

### Todo:
- Removal of EasyLoading plugin (?)
- Prevent problems caused by persisting BackupSettings. Which requires some changes on EnableBackupDialog(the dialog that opens up on clicking backup indicator on home page).
Let's say you picked Google Drive but decided not to restore and create a new account using Let's Breez.
The GDrive selection is persisted and clicking Backup Now through EnableBackupDialog forces you to login with GDrive. Unless you have an active backup, the Backup Now should behave similarly to selecting provider on Security & Backup settings page.

File: lib/routes/security_pin/remote_server_auth/remote_server_auth.dart [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+                        ),
+                      ),
+                    );
+                    return false;
+                  },
+                );
+              }
+            }
+            if (!uri.host.endsWith('.onion') && uri.scheme == 'http') {
+              promptAreYouSure(
```
Comment (ubbabeck) [inline lib/routes/security_pin/remote_server_auth/remote_server_auth.dart:172]: Tested with remote server over http and it breaks. Seems like it attempts to call testAuth before we allow connection.

Comment (erdemyerebasmaz) [reply lib/routes/security_pin/remote_server_auth/remote_server_auth.dart]: Resolved. 
- Refactor RemoteServerAuthPage auth logic 1e4fb77

**Comment by erdemyerebasmaz** [comment]
This issue is

> Let's say you picked Google Drive but decided not to restore and create a new account using Let's Breez.
> The GDrive selection is persisted and clicking Backup Now through EnableBackupDialog forces you to login with GDrive. Unless you have an active backup, the Backup Now should behave similarly to selecting provider on Security & Backup settings page.

addressed with:

- Reset backup settings before registering via Let's Breez a12e657
- Re-add userProfileBloc to InitialWalkthroughPage params 30325b1
- Improve error handling of Let's Breez flow dc91c22
- Set iCloud as default BackupProvider for iOS b815317
  - Revert always update backup settings changes 78c9eb8
  - Handle null backupProvider inside _updateBackupProvider fn 702cf3a
- Sign out of GDrive if settings are being reset 10b4e47

**Pull request merged**